File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
private:
void backtrack(int n, vector<string>&ans, string current,int openN,int closeN)
{
    if (openN==closeN && openN==n)
    {ans.push_back(current);
       return ;
        }

    if (openN<n){
        current.push_back('(');
        backtrack(n,ans,current,openN+1,closeN);
        current.pop_back();
    }

    if(closeN<openN){
        current.push_back(')');
        backtrack(n,ans,current,openN,closeN+1);
        current.pop_back();
    }

}
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string current;
        backtrack(n,ans,current,0,0);
        return ans;   
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        #only add open paranthesis if open<n
        #only add a closing paranthesis if closed< open
        #valid if open==close==n

        current=[]
        ans=[]

        def backtrack(openN,closeN):
            if openN == closeN==n:
                ans.append("".join(current))
                return

            
            if openN<n:
                current.append("(")
                backtrack(openN+1,closeN)
                current.pop()

            
            if closeN<openN:
                current.append(")")
                backtrack(openN,closeN+1)
                current.pop()

            
        backtrack(0,0)
        return ans


        